Домашнее задание, 3'я неделя

Muhamob opened the pull request.

Muhamob committed:
Add hometask-1

Muhamob committed:
Разделил класс EventList на конфигурационный код и класс для проверки всех событий

Muhamob committed:
Вынес из метода обработки событий метод обработки одного события, создал интерфейс для различных источников событий, сделал источник случайных событий

Muhamob committed:
Реализовал паттерн композитор, однако сериализация перестала работать из-за назначения комнаты девайсу

Muhamob committed:
Добавил поле smartDevices в smartHome

Muhamob committed:
Добавил метод execute в интерфейс HomeComponent, добавил интерфейс HomeComponentAction

Muhamob committed:
Доделал задание 2

Muhamob committed:
Code cleanup, add documentation

Muhamob committed:
Delete redundant sout's

Muhamob committed:
Add tests

Muhamob committed:
Добавил классы, отвечающие за состояние сигнализации по шаблону Состояние

Muhamob committed:
Добавил классы, отвечающие за состояние сигнализации по шаблону Состояние

Muhamob committed:
Добавил тесты для объекта сигнализации и переходов между состояниями

Muhamob committed:
Добавил тесты для объекта сигнализации и переходов между состояниями

Muhamob committed:
Исправил логику работы метода invokeAlert, добавил getter для AlarmState

Muhamob committed:
Исправил логику работы метода invokeAlert, добавил getter для AlarmState

Muhamob committed:
Добавил обработчики событий активации/деактивациии сигнализации

Muhamob committed:
Добавил обработчики событий активации/деактивациии сигнализации

Muhamob committed:
Добавил тестов на обработчики событий сигнализации

Muhamob committed:
Добавил тестов на обработчики событий сигнализации

Muhamob committed:
Добавил декораторы

Muhamob committed:
Добавил декораторы

Muhamob committed:
Переделал конфигурационный код, #1

Muhamob committed:
Исправил проблему leaky abstraction из первого задания, #2 #1

Muhamob committed:
Убрал ненужный метод

Muhamob committed:
Refactor signature of Actions constructor

Muhamob committed:
Refactor signature of Actions constructor

Muhamob committed:
Fix: alarm activation/deactivation logic

Muhamob committed:
Fix: alarm activation/deactivation logic

Muhamob committed:
Fix: test naming

Muhamob committed:
Fix: test naming

Muhamob committed:
Выделил цикл while в отдельную responsibility, #1 #2

Muhamob committed:
Выделил цикл while в отдельную responsibility, cherry-pick из второй ветки, #3

a-ivanov reviewed (changes requested):
Поправьте сначала замечания по 1 и 2 дз.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/eventHandlers/EventList.java:
```diff
@@ -0,0 +1,33 @@
+            HomeComponentAction action = eventAction.getAction(event);
+
+            // Декоратор для событий, если сигнализация забила тревогу
+            if (smartHome.getSmartDevice(smartHome.getAlarmId()) != null) {
```
Нарушение SRP. С сигнализацией должен работать обработчик событий, сейчас эта функция внутри класса с циклом событий.

Muhamob replied:
Исправил в https://github.com/Muhamob/smart-home-2019/pull/3/commits/491efc4a48cf554adf1496e38c3b96cb1c6abc55. Сейчас обёртывание действия происходит во время создания этих действий

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/sources/EventSourceManger.java:
```diff
@@ -0,0 +1,24 @@
+
+        while (event != null) {
+            System.out.println("Got event: " + event);
+            EventList.run(smartHome, event);
```
Нарушение OCP. А если захотим вызывать обработчики по-другому, а не через EventList.run? Сделайте зависимостью.

Muhamob replied:
Испрваил в https://github.com/Muhamob/smart-home-2019/pull/3/commits/00b00bc73e92457907ade0f584ac2c1efe9a9036. Теперь, если возникает необходимость вызывать обаботчики по-другому, нужно создать соответствующий класс, который будет реализовывать интерфейс EventManeger

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/actions/ActionWhileAlarmActivatedDecorator.java:
```diff
@@ -0,0 +1,29 @@
+                    return false;
+                }
+            });
+            System.out.println("Sending sms");
```
Почему смс отправляется здесь и в ActionWhileAlarmAlertDecorator? Надо делать это где-то в одном месте, DRY принцип.

Muhamob replied:
Исправил в https://github.com/Muhamob/smart-home-2019/pull/3/commits/aa7dcef0b03dce7b59753a6048acf5c00d9f195e . Теперь существует отдельный SmartDevice SMSSender, который отпарвляет в случае срабатывания сигнализации смс

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/sources/EventManager.java:
```diff
@@ -0,0 +1,7 @@
+import ru.sbt.mipt.oop.homeStructure.SmartHome;
+
+public interface EventManager {
+    public void manage(SmartHome smartHome);
```
Плохое название метода, непонятно, что он делает, конкретизируйте.

Muhamob replied:
Исправил название в https://github.com/Muhamob/smart-home-2019/pull/3/commits/93ca51d873e03185818bdf1980847c3cbb6d54af . EventProcessor может быть звучит лучше чем EventManager

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/sources/EventSource.java:
```diff
@@ -0,0 +1,14 @@
+    3. Какие-то реальные источники событий
+ */
+public interface EventSource {
+    public SensorEvent getNextSensorEvent();
```
В интерфейсах все методы public по умолчанию.

Muhamob replied:
Не знал. Исправил в https://github.com/Muhamob/smart-home-2019/pull/3/commits/9daed6e180c330579d8614f9c2a327b4579cd557

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/homeStructure/Interior.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Override
+    public SmartDevice getSmartDevice(String id) {
```
Не надо делать методы с доступом ко внутреннему сотсоянию только для тестов. Уберите их и подобные и пользуйтесь только execute.

Muhamob replied:
Исправил в https://github.com/Muhamob/smart-home-2019/commit/aa7dcef0b03dce7b59753a6048acf5c00d9f195e . Добавил класс HomeUtils, который имеет методы для получения SmartDevice по id или классу

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/homeStructure/Interior.java:
```diff
@@ -0,0 +1,60 @@
+    @Override
+    public boolean contains(String id) {
+        for (Floor floor : floors) {
+            if (floor.contains(id)) return true;
```
Floor должен сам выполнять цикл обработки внутри сам без доступа к его внутренним объектам - в этом смысл 2го задания.

Muhamob replied:
Floor теперь нет

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/homeStructure/Floor.java:
```diff
@@ -0,0 +1,60 @@
+    @Override
+    public boolean contains(String id) {
+        for (Room room : rooms) {
+            if (room.contains(id)) return true;
```
Room не должна давать доступ к своим внутренностям, в этом смысл задания 2.

Muhamob replied:
Исправлено в https://github.com/Muhamob/smart-home-2019/pull/3/commits/47fb25258e563bbd383d2841fbdcf8f673de90cb

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/homeStructure/HomeComponent.java:
```diff
@@ -0,0 +1,13 @@
+import java.util.List;
+
+public interface HomeComponent {
+    public SmartDevice getSmartDevice(String id);
```
Уберите эти методы из интерфейса, мы работаем с внутренним итератором execute только, в этом суть 2го дз.

Muhamob replied:
Исправлено в https://github.com/Muhamob/smart-home-2019/pull/3/commits/47fb25258e563bbd383d2841fbdcf8f673de90cb

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/homeStructure/Floor.java:
```diff
@@ -0,0 +1,60 @@
+
+    @Override
+    public boolean contains(String id) {
+        for (Room room : rooms) {
```
Почему тогда Floor не дает выполнить действие над ним самим, а только над комнатами?

Muhamob replied:
Исправлено в https://github.com/Muhamob/smart-home-2019/pull/3/commits/47fb25258e563bbd383d2841fbdcf8f673de90cb

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/eventHandlers/ActionEnum.java:
```diff
@@ -0,0 +1,35 @@
+Конфигурационный код
+Чтобы добавить ещё одно действие нужно дописать ACTION_SMTH { @Override HomeComp.... };
+ */
+public enum ActionEnum {
```
А если мы захотим добавить несколько обработчиков на одно и то же событие? В текущем решении это сделать нельзя. Эту и исходную проблему можно решить путем переноса ответственности маппинга типа события из захардкоженного в enum'е в отдельный класс обработчика. Исправьте.

Muhamob replied:
Испрвлено в https://github.com/Muhamob/smart-home-2019/pull/3/commits/8c83a3afae4d9ef16e35e884d89773eb3920ce44

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/homeStructure/Exterior.java:
```diff
@@ -0,0 +1,45 @@
+        for (HomeComponent homeComponent : homeComponents) {
+            executed |= homeComponent.execute(action);
+        }
+
```
Опять же, с Exterior тоже можно совершать действия.

Muhamob replied:
Удалено

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/eventHandlers/EventList.java:
```diff
@@ -0,0 +1,33 @@
+
+            // Декоратор для событий, если сигнализация забила тревогу
+            if (smartHome.getSmartDevice(smartHome.getAlarmId()) != null) {
+                Alarm alarm = (Alarm) smartHome.getSmartDevice(smartHome.getAlarmId());
```
Уберите метод getSmartDevice, по 2му заданию мы отказываемся от раскрытия внутренностей объектов дома.

Muhamob replied:
Исправлено в https://github.com/Muhamob/smart-home-2019/commit/47fb25258e563bbd383d2841fbdcf8f673de90cb

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/eventHandlers/EventList.java:
```diff
@@ -0,0 +1,33 @@
+
+            }
+
+            smartHome.execute(action);
```
Запуск action над элементами дома - это отдельная ответственность, сейчас тут нарушение SRP.

a-ivanov replied:
Комментарий выше все еще актуален, EventList максимум запускает обработку событий над списком обработчиком, переделайте.

Muhamob replied:
Немного не поинмаю почему этот код нарушает SRP, может быть я назвал класс запутанно. Однако, в этом классе происходит только запуск всех action

a-ivanov replied:
Да, я не сразу понял из-за названия класса HomeComponentAction (думал - это имплементация самого действия из 2го дз). Лучше переименовать HomeComponentAction в EventProcessor, а EventProcessor в EventLoopRunner, имплементация может носить название ListLoopRunner.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/devices/AlarmState.java:
```diff
@@ -0,0 +1,35 @@
+/*
+Интерфейс, используемый в шаблоне Состояние
+ */
+public abstract class AlarmState {
```
Перечитайте 3е задание. В интерфейсе состояния должны присутствовать 3 метода на каждое действие с сигнализацией. Переделывайте.

Muhamob committed:
Убрал ненужные методы, которые раскрывают устройство дома. Исправил в соответствии с этим тесты.

Muhamob committed:
Переделал EventList таким образом, чтобы он теперь не использовал рефлексию.

Muhamob committed:
Убрал ненужные абстракции и классы

Muhamob committed:
Rename: HomeComponent -> Actionable

Muhamob committed:
Optimize imports

Muhamob committed:
Optimize imports

Muhamob committed:
Rebase hometask-2 onto this branch, and fixies after it

Muhamob committed:
Fix: new home serialization

Muhamob committed:
Merge origin and local

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/devices/AlarmActivated.java:
```diff
@@ -0,0 +1,7 @@
+package ru.sbt.mipt.oop.devices;
+
+public class AlarmActivated extends AlarmState {
```
Чем AlarmActivated отличается от AlarmActivated от AlarmAlert? Они же делают все одно и то же, перечитайте, как надо имплементировать паттерн состояние. Откажитесь от абстрактного класса, сделайте интерфейс состояния и имплементируйте три класса на каждое из состояний.

Muhamob replied:
Переделал в https://github.com/Muhamob/smart-home-2019/pull/3/commits/699a181cc640d26ec074f0f32ea738c498a12b20

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/eventHandlers/EventList.java:
```diff
@@ -0,0 +1,42 @@
+
+            // Трюк, чтобы можно было внутри лямбда выражения назначить
+            // переменную из внешней области видимости
+            Alarm alarm = AlarmUtils.getAlarm(smartHome);
```
Ненужный трюк, можно либо создать сигнализацию вне дома, можно в доме геттер объявить с сигнализацией.

Muhamob replied:
Переделал в https://github.com/Muhamob/smart-home-2019/commit/aa7dcef0b03dce7b59753a6048acf5c00d9f195e

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/eventHandlers/EventList.java:
```diff
@@ -0,0 +1,42 @@
+
+            if (alarm != null) {
+
+                if (alarm.isActivated()) {
```
Вся логика с сигнализацией и ее состоянием - внутри декоратора.

Muhamob replied:
Переделал в https://github.com/Muhamob/smart-home-2019/commit/491efc4a48cf554adf1496e38c3b96cb1c6abc55

Muhamob committed:
Fix: delete contains method from Actionable interface

Muhamob committed:
Fix: delete contains methdo from SmartDevice, forgot it in last commit

Muhamob committed:
Remove: redudant IOException in Application class

Muhamob committed:
Fix: typo in Main class

Muhamob committed:
Fix: OCP for different event managers

Now it is possible to use different event managers that can execute actions in some way. In order to do it create new class like SomethingEventManeger that implements EventManager interface and override method manage

Muhamob committed:
Fix: SRP in EventList. Now EventCollectionCreator is responsible for handling alarm states in actions

Muhamob committed:
Add: SMSSender, HomeUtils.

Muhamob committed:
Rename: EventManager -> EventProcessor

Muhamob committed:
Fix: remove redundant modifiers in interface

Muhamob committed:
Fix: refactor alert system to follow state template.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/devices/alarm/AlarmActivated.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Override
+    public void activate(String code) {
+        // Логично, что если кто-то попытается ещё раз активировать сигнализацию
```
Не логично, этого добавлять было не обязательно.

Muhamob replied:
Изменил в https://github.com/Muhamob/smart-home-2019/pull/3/commits/3bd98a1920012e6a136c15b73ea6d0e3baa5a066

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/utils/HomeUtils.java:
```diff
@@ -0,0 +1,38 @@
+        return smartDeviceTmp[0];
+    }
+
+    public static SmartDevice getSmartDevice(Actionable actionable, Class clazz) {
```
Зачем это все, если можно просто добавить геттер в доме на сигнализацию? Зачем так абстрактно?

Muhamob replied:
Я принял данное решение, так как иначе если делать геттер в доме, это значит, что в самом объекте дома необходимо создать новое поле с сигнализацией, что противоречит принципу dependency inversion(если я правильно понимаю), так как при добавлении всего лишь одного нового смарт-девайса, приходится переписывать класс, который, по идее, не должен страдать/изменятся от добавления нового смарт-девайса. Поэтому сигнализация в моём решении является обычным девайсом в том смысле, что он может находится где угодно, а соответственно, чтобы его использовать, его нужно найти в доме. Для этого и существует данная функция

a-ivanov replied:
Все объекты дома - доменные классы, их API известен изначально по ТЗ и не должен сильно меняться. В них можно легко добавить геттеры с конкретными объектами, это сильно упрощает API и работу с ними.

Muhamob replied:
fd57261

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/eventHandlers/EventCollectionCreator.java:
```diff
@@ -0,0 +1,54 @@
+    public List<HomeComponentAction> getActionList(SensorEvent event) {
+        List<HomeComponentAction> actions = new ArrayList<>();
+
+        for (HomeComponentAction action : getDefaultActionList(event)) {
```
Весь for блок - это нарушение OCP, в этом классе максимум создание списка обработчиков. Логика по запуску обработки - в обработчиках.

Muhamob replied:
В данном классе действия только создаются и максимум оборачиваются декоратором

a-ivanov replied:
Здесь написана часть логики декоратора, ее надо перенести в код декоратора. Сделайте один декоратор, зачем два узкоспециализированных класса, которые тяжело переиспользовать.

Muhamob replied:
e33d6c6

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/actions/ActionWhileAlarmActivatedDecorator.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Override
+    public boolean execute(Actionable homeComponent) {
+        boolean executed = wrappee.execute(homeComponent);
```
По заданию декоратор должен а) контролировать запуск дальнейших обработчиков в зависимости от статуса сигнализации и б) отправлять смс. Здесь непонятно, при каких условиях будет запущен wrappee. Я вижу, что ему дают запускаться всегда. Бизнес логика не ложна зависеть от конфигурации, переделайте.

Muhamob replied:
Да, wrappee будет запускаться всегда, однако дело в том, что не всегда действия оборачиваются данным декоратором. В классе EventCollectionCreator определяется будет ли действие обёрнуто в данный декоратор или нет в соответствии с бизнес-логикой.

a-ivanov replied:
У вас логика декоратора разнесена на два класса + конфигурационный класс, который "настраивает" декораторы. Должно быть все в одном классе декоратора. Здесь нарушается OCP, SRP (другие классы делают не свою работу). Переделайте.

Muhamob replied:
e33d6c6

Muhamob committed:
Fix: logic

Muhamob committed:
Fix: add alarm to SmartHome fields and getter to it.

Muhamob committed:
Refactor code to use only one alarm decorator

Muhamob committed:
Delete: redundant comments

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/actions/AlarmDecorator.java:
```diff
@@ -0,0 +1,58 @@
+
+        if (alarm != null) {
+            if (alarm.getAlarmState().getClass().equals(AlarmActivated.class)) {
+                if (wrappee.getClass() != AlarmAction.class) {
```
Проверка класса wrappee - это нарушение OCP, вам должно быть достаточно проверить состояние сигнализации и тип события. Так вы привязываетесь к конкретному классу обработчика, переделайте.

Muhamob replied:
Поправил в d1693d3

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/actions/AlarmDecorator.java:
```diff
@@ -0,0 +1,58 @@
+
+        if (alarm != null) {
+            if (alarm.getAlarmState().getClass().equals(AlarmActivated.class)) {
+                if (wrappee.getClass() != AlarmAction.class) {
```
Проверка класса wrappee - это нарушение OCP, вам должно быть достаточно проверить состояние сигнализации и тип события. Так вы привязываетесь к конкретному классу обработчика, переделайте.

Muhamob committed:
Update: fix AlarmDecorator to follow OCP principle

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/actions/AlarmDecorator.java:
```diff
@@ -0,0 +1,63 @@
+            if (alarm.getAlarmState().getClass().equals(AlarmActivated.class)) {
+//                if (wrappee.getClass() != AlarmAction.class) {
+                if (event.getType() != SensorEventType.ALARM_DEACTIVATE) {
+                    result = executeWhileActivated(actionable);
```
Ну так а в чем логика? Тут если AlarmActivated, то любое событие кроме ALARM_DEACTIVATE будет обработано. Смотрите задание 3, там необходимо прерывать обработку событий при активной сигнализации.

Muhamob replied:
Нет, в данном случае вызывается метод executeWhileAlarmActivated, который переключает сигнализацию в режим тревоги и отправляет смс

Muhamob replied:
Ошибся, не заметил, что там действительно исполняется действие в первой строчке. Исправил в 368bde6

Muhamob committed:
Fix logic of alarm activation/deactivation
